Handle IAnonymousObjectCreationOperation in CFG

mavasani opened the pull request.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -5031,6 +5033,65 @@ public override IOperation VisitMemberInitializer(IMemberInitializerOperation op
+                initializerBuilder.Add(GetCaptureReference(captureId, initializer));
+
+                // For VB, previously initialized properties can be reference in subsequent initializers.
+                // We store the capture Id for the property fur such property references.
```
>fur [](start = 60, length = 3)

Typo?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -5031,6 +5033,65 @@ public override IOperation VisitMemberInitializer(IMemberInitializerOperation op
+            void visitAndCaptureInitializer(IPropertySymbol initializedProperty, IOperation initializer)
+            {
+                int captureId = VisitAndCapture(initializer);
+                initializerBuilder.Add(GetCaptureReference(captureId, initializer));
```
>GetCaptureReference(captureId, initializer) [](start = 39, length = 43)

I think we should preserve the assignment if original initializer was one.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IAnonymousObjectCreationOperation.vb:
```diff
@@ -0,0 +1,1104 @@
+                  Operand: 
+                    IAnonymousObjectCreationOperation (OperationKind.AnonymousObjectCreation, Type: <anonymous type: Key a As System.Int32, Key b As System.Int32>) (Syntax: 'New With {  ... ey .b = j }')
+                      Initializers(2):
+                          IFlowCaptureReferenceOperation: 1 (OperationKind.FlowCaptureReference, Type: System.Int32, IsImplicit) (Syntax: 'i')
```
>IFlowCaptureReferenceOperation: 1 (OperationKind.FlowCaptureReference, Type: System.Int32, IsImplicit) (Syntax: 'i') [](start = 26, length = 116)

Was this an assignment in the original tree? I think we should keep it an assignment and this node should be its value part.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IAnonymousObjectCreationOperation.vb:
```diff
@@ -0,0 +1,1104 @@
+
+Class C
+    Sub M(p As Object, a As List(Of Integer), b As List(Of String))'BIND:"Sub M(p As Object, a As List(Of Integer), b As List(Of String))"
+        p = From x In a From y In b Select New With {x, y}
```
>New With {x, y} [](start = 43, length = 15)

I do not think scenarios with explicit Anonymous Object creation inside a query are particularly interesting. What is interesting is the Anonymous Object creation generated by the compiler implicitly. For example, for a query like ```From x In a From y In b ```. Please adjust test(s) accordingly

mavasani replied:
These scenarios did not generate an anonymous object creation, so I added the explicit creation. I have changed the tests and added a comment that the baselines are affected by [https://github.com/dotnet/roslyn/issues/25268](https://github.com/dotnet/roslyn/issues/25268)

---
In reply to: [192579293](https://github.com/dotnet/roslyn/pull/27370#discussion_r192579293) [](ancestors = 192579293)

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IAnonymousObjectCreationOperation.vb:
```diff
@@ -0,0 +1,1104 @@
+
+Class C
+    Sub M(p As Object, a As List(Of Integer))'BIND:"Sub M(p As Object, a As List(Of Integer))"
+        p = From x In a Let y = x Select New With {x, y}
```
>From x In a Let y = x Select New With {x, y} [](start = 12, length = 44)

Similarly, the interesting query would be ```From x In a Let y = x```

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -5031,6 +5033,65 @@ public override IOperation VisitMemberInitializer(IMemberInitializerOperation op
+
+                // For VB, previously initialized properties can be reference in subsequent initializers.
+                // We store the capture Id for the property fur such property references.
+                // Note that for error cases with duplicate property names, all the duplicate the property symbols are treated equals.
```
>all the duplicate the property symbols are treated equals. [](start = 76, length = 58)

It feels like the comment might need an adjustment

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -5461,6 +5522,17 @@ public override IOperation VisitMethodReference(IMethodReferenceOperation operat
+            if (operation.Instance == null &&
+                operation.Property.ContainingType.IsAnonymousType &&
+                operation.Property.ContainingType == _currentImplicitInstance.AnonymousType &&
+                _currentImplicitInstance.AnonymousTypePropertyCaptureIds.TryGetValue(operation.Property, out int captureId))
```
>_currentImplicitInstance.AnonymousTypePropertyCaptureIds.TryGetValue(operation.Property, out int captureId) [](start = 16, length = 107)

Should we create an invalid operation when we don't yet have a value for this reference instead of falling back to the default rewrite?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IAnonymousObjectCreationOperation.vb:
```diff
@@ -0,0 +1,1104 @@
+
+            VerifyFlowGraphAndDiagnosticsForTest(Of MethodBlockSyntax)(source, expectedFlowGraph, expectedDiagnostics)
+        End Sub
+    End Class
```
>End Class [](start = 4, length = 9)

It would be good to have a test for a scenario like ```new with {.a = .b, .b = .b}```

mavasani replied:
Added `AnonymousObjectCreation_NoControlFlow_08`

---
In reply to: [192579691](https://github.com/dotnet/roslyn/pull/27370#discussion_r192579691) [](ancestors = 192579691)

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IAnonymousObjectCreationOperation.vb:
```diff
@@ -0,0 +1,1104 @@
+          Value: 
+            ISimpleAssignmentOperation (OperationKind.SimpleAssignment, Type: System.Int32) (Syntax: 'i')
+              Left: 
+                IPropertyReferenceOperation: Property <anonymous type: i As System.Int32, b As System.Int32>.b As System.Int32 (OperationKind.PropertyReference, Type: System.Int32) (Syntax: 'b')
```
>IPropertyReferenceOperation: Property <anonymous type: i As System.Int32, b As System.Int32>.b As System.Int32 (OperationKind.PropertyReference, Type: System.Int32) (Syntax: 'b') [](start = 16, length = 178)

Should we create an invalid operation instead?

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IAnonymousObjectCreationOperation.cs:
```diff
@@ -0,0 +1,1126 @@
+          Value: 
+            ISimpleAssignmentOperation (OperationKind.SimpleAssignment, Type: System.Int32) (Syntax: 'a')
+              Left: 
+                IPropertyReferenceOperation: System.Int32 <anonymous type: System.Int32 a, System.Int32 b>.b { get; } (OperationKind.PropertyReference, Type: System.Int32) (Syntax: 'b')
```
>IPropertyReferenceOperation: System.Int32 <anonymous type: System.Int32 a, System.Int32 b>.b { get; } (OperationKind.PropertyReference, Type: System.Int32) (Syntax: 'b') [](start = 16, length = 169)

Should we create an invalid operation instead?

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IAnonymousObjectCreationOperation.cs:
```diff
@@ -0,0 +1,1126 @@
+    }/*</bind>*/
+}
+";
+            // Affected by https://github.com/dotnet/roslyn/issues/27338
```
>// Affected by https://github.com/dotnet/roslyn/issues/27338 [](start = 12, length = 60)

Is this comment accurate?

AlekseyTs commented:
Done with review pass (iteration 1)

mavasani commented:
@AlekseyTs addressed feedback

---
In reply to: [394143466](https://github.com/dotnet/roslyn/pull/27370#issuecomment-394143466) [](ancestors = 394143466)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -5031,6 +5033,71 @@ public override IOperation VisitMemberInitializer(IMemberInitializerOperation op
+                            IOperation visitedTarget = new PropertyReferenceExpression(initializedProperty, propertyReference.Instance, ImmutableArray<IArgumentOperation>.Empty,
+                                semanticModel: null, propertyReference.Syntax, propertyReference.Type, propertyReference.ConstantValue, IsImplicit(propertyReference));
+                            IOperation visitedValue = visitAndCaptureInitializer(initializedProperty, simpleAssignment.Value);
+                            var visitedAssignment = new SimpleAssignmentExpression(visitedTarget, isRef: false, visitedValue,
```
>isRef: false [](start = 98, length = 12)

Please use value from the original node.

jcouv commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.ImplicitInstanceInfo.cs:
```diff
@@ -0,0 +1,54 @@
+{
+    internal partial class ControlFlowGraphBuilder
+    {
+        /// <summary>
```
Nit: the comment being duplicated makes me think one of the two will go out of date.
Consider phrasing so that one the master (comprehensive) and the other just a summary (with list of examples, but clearly non-comprehensive).

jcouv commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.ImplicitInstanceInfo.cs:
```diff
@@ -0,0 +1,54 @@
+                AnonymousTypePropertyCaptureIds = null;
+            }
+
+            public ImplicitInstanceInfo(ITypeSymbol currentInitializedAnonymousType)
```
>ITypeSymbol [](start = 40, length = 11)

Should this constructor use an `INamedTypeSymbol` parameter instead? (ie. tightening the API contract)

jcouv commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IAnonymousObjectCreationOperation.cs:
```diff
@@ -0,0 +1,1370 @@
+                            Left: 
+                              IPropertyReferenceOperation: System.Int32 <anonymous type: System.Int32 a, System.Int32 b>.a { get; } (OperationKind.PropertyReference, Type: System.Int32) (Syntax: 'a')
+                                Instance Receiver: 
+                                  null
```
>null [](start = 34, length = 4)

This `null` receiver is indeed surprising. But I see that you have a PROTOTYPE comment for that already. Thanks

mavasani commented:
@dotnet-bot retest windows_debug_vs-integration_prtest please

mavasani commented:
@dotnet-bot retest windows_debug_spanish_unit32_prtest please

mavasani commented:
@dotnet-bot retest windows_debug_unit64_prtest  please

333fred commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.ImplicitInstanceInfo.cs:
```diff
@@ -0,0 +1,54 @@
+            /// <summary>
+            /// Holds the current object instance being initialized if we're visiting an object initializer.
+            /// </summary>
+            public IOperation Object { get; }
```
Nit: perhaps InitializedObject or ImplicitInstance

333fred commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -4949,6 +4950,71 @@ public override IOperation VisitMemberInitializer(IMemberInitializerOperation op
+            var properties = operation.Type.GetMembers().OfType<IPropertySymbol>().ToImmutableArray();
+            Debug.Assert(properties.Length == operation.Initializers.Length);
+
+            SpillEvalStack();
```
Is it possible to have an anonymous creation without initializers? In that case, should we not spill the eval stack?

mavasani replied:
It is allowed in C#, generates `BC36574: Anonymous type must contain at least one member.` in VB. I have added an early bailout for no initializers and also added unit tests.

---
In reply to: [192900607](https://github.com/dotnet/roslyn/pull/27370#discussion_r192900607) [](ancestors = 192900607)

333fred commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -4949,6 +4950,71 @@ public override IOperation VisitMemberInitializer(IMemberInitializerOperation op
+            {
+                int captureId = VisitAndCapture(initializer);
+                
+                // For VB, previously initialized properties can be reference in subsequent initializers.
```
>reference [](start = 68, length = 9)

Nit: typo, `reference`->`referenced`

333fred commented:
@mavasani test failures appear legitimate.

333fred commented:
Done review (Iteration 4)

mavasani commented:
@dotnet-bot retest windows_debug_vs-integration_prtest please

mavasani commented:
@dotnet-bot retest windows_release_vs-integration_prtest please

mavasani commented:
@333fred Can you take another look?

mavasani commented:
@dotnet-bot retest windows_debug_vs-integration_prtest please

mavasani commented:
Merging as @AlekseyTs has some more changes in pipeline that depend on this change. @333fred, will address any more feedback with a separate PR.
